### PR TITLE
[ Misc ] `fbgemm` checkpoints

### PR DESCRIPTION
**Pull request by robertgshaw2-redhat**
SUMMARY:
* support `fbgemm` checkpoints from https://github.com/huggingface/transformers/pull/32047/files using our existing cutlass kernels
* note: updated with latest state that removes `activation_scale_ub` from the state dict and instead uses the config.json

```python
from vllm import LLM
model = LLM("nm-testing/Meta-Llama-3-8B-Instruct-FBGEMM-nonuniform") 
```

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Review by mgoin — APPROVED** [review]
Nice work keeping it clean, LGTM

File: vllm/model_executor/layers/quantization/fbgemm_fp8.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+    def apply(self,
+              layer: torch.nn.Module,
+              x: torch.Tensor,
+              bias: Optional[torch.Tensor] = None) -> torch.Tensor:
+
+        return apply_fp8_linear(input=x,
+                                weight=layer.weight,
+                                weight_scale=layer.weight_scale,
+                                input_scale=None,
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/fbgemm_fp8.py:113]: Might be worth noting whether this will trigger dynamic per-tensor or per-token scales

Comment (robertgshaw2-neuralmagic) [reply vllm/model_executor/layers/quantization/fbgemm_fp8.py:113]: We only have per tensor at current state. Will update comment in dyn per token pr

File: vllm/model_executor/layers/quantization/fbgemm_fp8.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    def process_weights_after_loading(self, layer: Module) -> None:
+        weight = layer.weight
+        layer.weight = Parameter(weight.t(), requires_grad=False)
+
+    def apply(self,
+              layer: torch.nn.Module,
+              x: torch.Tensor,
+              bias: Optional[torch.Tensor] = None) -> torch.Tensor:
+
+        return apply_fp8_linear(input=x,
```
Comment (vlasenkoalexey) [inline vllm/model_executor/layers/quantization/fbgemm_fp8.py:110]: I don't think that apply_fp8_linear would work here. Using fbgemm it would look like:

            xq, x_scale = torch.ops.fbgemm.quantize_fp8_per_row(
                x, num_tokens, self.activation_scale_ub
            )
            y = torch.ops.fbgemm.f8f8bf16_rowwise(
                xq, layer.weight, x_scale, layer.weight_scale, use_fast_accum=True
            )

Particularly input_scale=None is most likely wrong, here is a reference implementation for quantize_fp8_per_row

def fp8_row_quantize_ref(x: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
    # Quantize an input tensor and return the fp8 tensor and its inverse scale.
    x_row_max = torch.max(torch.abs(x), dim=1).values
    max_scaling_factor = E4M3_MAX_POS * 512.0  # Match kernel logics
    scale = torch.Tensor(E4M3_MAX_POS / x_row_max).clamp(max=max_scaling_factor)
    xq = (x * scale.unsqueeze(1)).to(fp8_e4m3)
    return xq, scale.reciprocal().to(torch.float32)

Comment (robertgshaw2-neuralmagic) [reply vllm/model_executor/layers/quantization/fbgemm_fp8.py:110]: yup, we are updating the quant kernel right now to use the `activation_scale_ub`

Comment (robertgshaw2-neuralmagic) [reply vllm/model_executor/layers/quantization/fbgemm_fp8.py:110]: @varun-sundar-rabindranath

Comment (robertgshaw2-neuralmagic) [reply vllm/model_executor/layers/quantization/fbgemm_fp8.py:110]: in combo with https://github.com/vllm-project/vllm/pull/6547/files

which enables per token scales

Comment (vlasenkoalexey) [reply vllm/model_executor/layers/quantization/fbgemm_fp8.py:110]: Besides both torch._scaled_mm and ops.scaled_fp8_quant expect scale to be scalar

Comment (robertgshaw2-neuralmagic) [reply vllm/model_executor/layers/quantization/fbgemm_fp8.py:110]: Then it will match. `input_scale=None` in this function signals to use dynamic quantization rather than static

Comment (robertgshaw2-neuralmagic) [reply vllm/model_executor/layers/quantization/fbgemm_fp8.py:110]: #6547 extends `ops.scaled_fp8_quant` to accepts per token scales (vector of scales)

Comment (robertgshaw2-neuralmagic) [reply vllm/model_executor/layers/quantization/fbgemm_fp8.py:110]: `torch._scaled_mm` will not be used.

`cutlass_scaled_mm` accepts per channel weights and per row activations

Comment (robertgshaw2-neuralmagic) [reply vllm/model_executor/layers/quantization/fbgemm_fp8.py:110]: I pulled in #6547 to this PR so you can see

Comment (robertgshaw2-neuralmagic) [reply vllm/model_executor/layers/quantization/fbgemm_fp8.py:110]: Adds https://github.com/vllm-project/vllm/pull/6593/ adds the ub

Comment (robertgshaw2-neuralmagic) [reply vllm/model_executor/layers/quantization/fbgemm_fp8.py:110]: Okay, this PR currently now has the 
- `scale_ub`
- uses dynamic per token activation scales

File: vllm/model_executor/layers/quantization/fbgemm_fp8.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+
+    def _is_layer_skipped(self, prefix: str) -> bool:
+        # prefix: model.layers.0.self_attn.q_proj
+        # proj_name: q_proj
+        proj_name = prefix.split(".")[-1]
+        if proj_name in _FUSED_LAYER_NAME_MAPPING:
+            shard_prefixes = [
+                prefix.replace(proj_name, shard_proj_name)
+                for shard_proj_name in _FUSED_LAYER_NAME_MAPPING[proj_name]
+            ]
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/fbgemm_fp8.py:62]: Nice

**Review by mgoin — APPROVED** [review]
LGTM!

File: .buildkite/lm-eval-harness/configs/Meta-Llama-3-8B-Instruct-Channelwise-compressed-tensors.yaml [hunk 4]
```diff
@@ -4,8 +4,8 @@ tasks:
 - name: "gsm8k"
```
Comment (mgoin) [inline .buildkite/lm-eval-harness/configs/Meta-Llama-3-8B-Instruct-Channelwise-compressed-tensors.yaml:4]: nit: it would be nice to rename this config to `Meta-Llama-3-8B-Instruct-W8A8-FP8-Channelwise-compressed-tensors.yaml`
